File: src/main/java/com/back/domain/chat/service/ChatService.java
package com.back.domain.chat.service;

import com.back.domain.auction.auction.entity.Auction;
import com.back.domain.auction.auction.repository.AuctionRepository;
import com.back.domain.chat.dto.ChatDto;
import com.back.domain.chat.entity.Chat;
import com.back.domain.chat.entity.ChatRoom;
import com.back.domain.chat.entity.ChatRoomType;
import com.back.domain.chat.repository.ChatRepository;
import com.back.domain.chat.repository.ChatRoomRepository;
import com.back.domain.member.member.entity.Member;
import com.back.domain.member.member.repository.MemberRepository;
import com.back.domain.post.post.entity.Post;
import com.back.domain.post.post.entity.PostStatus;
import com.back.domain.post.post.repository.PostRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class ChatService {

    private final ChatRepository chatRepository;
    private final ChatRoomRepository chatRoomRepository;
    private final PostRepository postRepository;
    private final AuctionRepository auctionRepository;
    private final MemberRepository memberRepository;

    /**
     * 채팅방 생성
     * @param itemId 상품 ID (Post ID or Auction ID)
     * @param txType POST or AUCTION
     * @param buyerApiKey 구매자 API Key
     * @return roomId (UUID)
     */
    @Transactional
    public String createChatRoom(int itemId, String txType, String buyerApiKey) {
        ChatRoomType type = ChatRoomType.valueOf(txType.toUpperCase());

        Member buyer = memberRepository.findByApiKey(buyerApiKey)
                .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 회원입니다."));

        if (type == ChatRoomType.POST) {
            Post post = postRepository.findById(itemId)
                    .orElseThrow(() -> new IllegalArgumentException("해당 게시글이 존재하지 않습니다."));

            // SALE 상태가 아니면 채팅방 생성 불가능
            if (post.getStatus() != PostStatus.SALE) {
                throw new IllegalStateException("판매 중인 상품이 아니므로 채팅을 시작할 수 없습니다.");
            }

            // 셀프 채팅 방지
            if (post.getSeller().getApiKey().equals(buyerApiKey)) {
                throw new IllegalArgumentException("본인의 상품에는 채팅을 개설할 수 없습니다.");
            }

            // 이미 존재하는 방인지 확인 후 없으면 생성
            return chatRoomRepository.findByPostAndBuyerId(post, buyerApiKey)
                    .map(ChatRoom::getRoomId)
                    .orElseGet(() -> {
                        ChatRoom room = ChatRoom.createForPost(post, buyer);
                        chatRoomRepository.save(room);
                        return room.getRoomId();
                    });

        } else if (type == ChatRoomType.AUCTION) {
            Auction auction = auctionRepository.findById(itemId)
                    .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 경매입니다."));

            return chatRoomRepository.findByAuctionAndBuyerId(auction, buyerApiKey)
                    .map(ChatRoom::getRoomId)
                    .orElseGet(() -> {
                        ChatRoom room = ChatRoom.createForAuction(auction, buyer);
                        chatRoomRepository.save(room);
                        return room.getRoomId();
                    });

        } else {
            throw new IllegalArgumentException("채팅을 개설할 수 없습니다.");
        }
    }

    @Transactional
    public void saveMessage(ChatDto chatDto) {
        // 방 존재 여부 검증
        ChatRoom room = chatRoomRepository.findByRoomId(chatDto.roomId())
                        .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 채팅방입니다."));

        Chat chatMessage = Chat.builder()
                .chatRoom(room)
                .sender(chatDto.sender())
                .message(chatDto.message())
                .isRead(false)
                .build();

        chatRepository.save(chatMessage);
    }

    @Transactional
    public List<ChatDto> getMessages(String roomId, Integer lastChatId, String readerName) {
        if (readerName != null) {
            chatRepository.markMessagesAsRead(roomId, readerName);
        }

        List<Chat> chats;
        if (lastChatId == null || lastChatId <= 0) {
            chats = chatRepository.findAllByRoomIdOrderByCreateDateAsc(roomId);
        } else {
            chats = chatRepository.findByRoomIdAndIdGreaterThanOrderByCreateDateAsc(roomId, lastChatId);
        }

        return chats.stream()
                .map(this::chatDto)
                .collect(Collectors.toList());
    }

    public List<ChatDto> getChatList() {
        return chatRepository.findAllLatestMessages().stream()
                .map(this::chatDto)
                .collect(Collectors.toList());
    }

    private ChatDto chatDto(Chat chat) {
        ChatRoom room = chat.getChatRoom();

        Integer itemId = null;
        if (room.getTxType() == ChatRoomType.POST) {
            itemId = (room.getPost() != null) ? room.getPost().getId() : null;
        } else if (room.getTxType() == ChatRoomType.AUCTION) {
            itemId = (room.getAuction() != null) ? room.getAuction().getId() : null;
        }

        return new ChatDto(
                chat.getId(),
                itemId,
                room.getRoomId(),
                chat.getSender(),
                chat.getMessage(),
                chat.getCreateDate(),
                chat.isRead());
    }
}

File: src/main/java/com/back/domain/chat/chat/service/ChatService.kt
package com.back.domain.chat.chat.service

import com.back.domain.chat.chat.dto.request.ChatMessageRequest
import com.back.domain.chat.chat.dto.response.ChatIdResponse
import com.back.domain.chat.chat.dto.response.ChatNotification
import com.back.domain.chat.chat.dto.response.ChatResponse
import com.back.domain.chat.chat.dto.response.ChatRoomIdResponse
import com.back.domain.chat.chat.dto.response.ChatRoomListResponse
import com.back.domain.chat.chat.entity.Chat
import com.back.domain.chat.chat.entity.ChatRoom
import com.back.domain.chat.chat.entity.ChatRoomType
import com.back.domain.chat.chat.service.event.ChatMessageCommittedEvent
import com.back.domain.chat.chat.service.event.ChatRoomReadEvent
import com.back.domain.chat.chat.service.port.ChatItemPort
import com.back.domain.chat.chat.service.port.ChatMemberInfo
import com.back.domain.chat.chat.service.port.ChatMemberPort
import com.back.domain.chat.chat.service.port.ChatMediaPort
import com.back.domain.chat.chat.service.port.ChatPersistencePort
import com.back.domain.chat.chat.service.port.ChatPublishPort
import com.back.domain.chat.chat.service.port.ChatUploadFile
import com.back.domain.chat.chat.service.port.ChatUseCase
import com.back.global.exception.ServiceException
import com.back.global.rq.Rq
import com.back.global.rsData.RsData
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.ApplicationEventPublisher
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import java.util.Locale
import java.util.concurrent.ConcurrentHashMap

@Service
@Transactional(readOnly = true)
class ChatService(
    private val chatPersistencePort: ChatPersistencePort,
    private val chatItemPort: ChatItemPort,
    private val chatMemberPort: ChatMemberPort,
    private val chatMediaPort: ChatMediaPort,
    private val chatPublishPort: ChatPublishPort,
    private val rq: Rq,
    private val eventPublisher: ApplicationEventPublisher,
    @Value("\${chat.events.async-enabled:true}")
    private val asyncEventsEnabled: Boolean,
) : ChatUseCase {
    private val memberCache = ConcurrentHashMap<Int, CachedCurrentMember>()

    @Transactional
    override fun createChatRoom(itemId: Int, txType: String): RsData<ChatRoomIdResponse> {
        val type = parseTxType(txType)
        val buyer = currentMemberFromDb()

        val createdRoom = when (type) {
            ChatRoomType.POST -> {
                val item = chatItemPort.getPostItemOrThrow(itemId)
                ensureNotSelfTrade(item.sellerApiKey, buyer.apiKey)

                val existingRoom = chatPersistencePort.findExistingRoom(type, itemId, buyer.apiKey)
                val isNew = existingRoom == null
                val room = existingRoom ?: chatPersistencePort.saveRoom(
                    ChatRoom.createForPost(
                        itemId = item.itemId,
                        itemName = item.itemName,
                        itemPrice = item.itemPrice,
                        itemImageUrl = item.itemImageUrl,
                        sellerApiKey = item.sellerApiKey,
                        buyerApiKey = buyer.apiKey,
                    ),
                )
                CreatedRoom(room = room, sellerApiKey = item.sellerApiKey, isNew = isNew)
            }

            ChatRoomType.AUCTION -> {
                val item = chatItemPort.getAuctionItemOrThrow(itemId)
                ensureNotSelfTrade(item.sellerApiKey, buyer.apiKey)

                val existingRoom = chatPersistencePort.findExistingRoom(type, itemId, buyer.apiKey)
                val isNew = existingRoom == null
                val room = existingRoom ?: chatPersistencePort.saveRoom(
                    ChatRoom.createForAuction(
                        itemId = item.itemId,
                        itemName = item.itemName,
                        itemPrice = item.itemPrice,
                        itemImageUrl = item.itemImageUrl,
                        sellerApiKey = item.sellerApiKey,
                        buyerApiKey = buyer.apiKey,
                    ),
                )
                CreatedRoom(room = room, sellerApiKey = item.sellerApiKey, isNew = isNew)
            }
        }

        if (createdRoom.isNew) {
            val seller = chatMemberPort.findMemberByApiKey(createdRoom.sellerApiKey)
                ?: throw ServiceException("404-1", "존재하지 않는 회원입니다.")

            sendUserNotification(
                type = "NEW_ROOM",
                room = createdRoom.room,
                recipient = seller,
                opponent = buyer,
                message = null,
                messageDate = LocalDateTime.now(),
            )
        }

        return RsData("200-1", "채팅방에 입장했습니다.", ChatRoomIdResponse(createdRoom.room.roomId))
    }

    @Transactional
    override fun saveMessage(req: ChatMessageRequest): RsData<ChatIdResponse> {
        val roomId = req.roomId ?: throw ServiceException("400-1", "채팅방 ID는 필수입니다.")
        val room = findActiveRoom(roomId)

        val sender = currentMemberFromDb()
        requireChatParticipant(room, sender.apiKey)

        val chatMessage = Chat(
            chatRoom = room,
            senderId = sender.id,
            message = req.message,
            read = false,
        )

        val uploadFiles = req.images.orEmpty().map {
            ChatUploadFile(
                filename = it.originalFilename,
                contentType = it.contentType,
                bytes = it.bytes,
            )
        }
        chatMediaPort.saveChatImages(chatMessage, uploadFiles)
        val persistedMessage = chatPersistencePort.saveChat(chatMessage)

        val chatResponse = ChatResponse.from(persistedMessage, sender.profileImageUrl)
        if (asyncEventsEnabled) {
            eventPublisher.publishEvent(
                ChatMessageCommittedEvent(
                    roomId = roomId,
                    senderId = sender.id,
                    senderApiKey = sender.apiKey,
                    senderNickname = sender.nickname,
                    senderProfileImageUrl = sender.profileImageUrl,
                    txType = room.txType,
                    itemId = room.itemId,
                    itemName = room.itemName,
                    itemImageUrl = room.itemImageUrl,
                    itemPrice = room.itemPrice,
                    sellerApiKey = room.sellerApiKey,
                    buyerApiKey = room.buyerApiKey,
                    message = req.message,
                    messageDate = persistedMessage.createDate,
                    roomMessagePayload = chatResponse,
                ),
            )
        } else {
            val opponentApiKey = if (room.sellerApiKey == sender.apiKey) room.buyerApiKey else room.sellerApiKey
            runMessagingSafely(
                onSuccess = { log.info("메시지 브로드캐스팅 성공 - RoomID: {}, MessageID: {}", roomId, persistedMessage.id) },
                onFailure = { e -> log.error("WebSocket 전송 실패 - RoomID: {}, Error: {}", roomId, e.message) },
            ) {
                chatPublishPort.publishRoomMessage(roomId, chatResponse)
            }

            chatMemberPort.findMemberByApiKey(opponentApiKey)?.let { opponent ->
                sendUserNotification(
                    type = "NEW_MESSAGE",
                    room = room,
                    recipient = opponent,
                    opponent = sender,
                    message = req.message,
                    messageDate = persistedMessage.createDate,
                )
            }
        }

        return RsData("200-1", "메시지가 전송되었습니다.", ChatIdResponse(persistedMessage.id))
    }

    @Transactional
    override fun getMessages(roomId: String, lastChatId: Int?): RsData<List<ChatResponse>> {
        val room = findActiveRoom(roomId)

        val me = currentMemberFromDb()
        requireChatParticipant(room, me.apiKey)

        val chats = chatPersistencePort.findRecentChats(roomId, lastChatId)
        val unreadMessageIds = chats.asSequence()
            .filter { !it.read && it.senderId != me.id }
            .map { it.id }
            .toList()
        val unreadMessageIdSet = unreadMessageIds.toSet()

        val updatedCount = if (unreadMessageIds.isEmpty()) 0 else chatPersistencePort.markMessagesAsReadByIds(unreadMessageIds)
        if (updatedCount > 0) {
            if (asyncEventsEnabled) {
                eventPublisher.publishEvent(
                    ChatRoomReadEvent(
                        roomId = roomId,
                        readerId = me.id,
                        updatedCount = updatedCount,
                    ),
                )
            } else {
                val readNotification: Any = mapOf(
                    "readerId" to me.id,
                    "roomId" to roomId,
                )
                chatPublishPort.publishRoomRead(roomId, readNotification)
                log.debug("읽음 알림 전송 - RoomId: {}, ReaderId: {}, UpdatedCount: {}", roomId, me.id, updatedCount)
            }
        }

        val membersByApiKey = chatMemberPort.findMembersByApiKeys(setOf(room.sellerApiKey, room.buyerApiKey))
        val profileByMemberId = membersByApiKey.values.associateBy({ it.id }, { it.profileImageUrl })

        val responses = chats.asReversed().map { chat ->
            val senderProfile = profileByMemberId[chat.senderId]
            ChatResponse.from(
                chat = chat,
                roomId = room.roomId,
                itemId = room.itemId,
                senderProfileImageUrl = senderProfile,
                readOverride = chat.read || unreadMessageIdSet.contains(chat.id),
            )
        }

        return RsData("200-1", "메시지 조회 성공", responses)
    }

    override fun getChatList(): RsData<List<ChatRoomListResponse>> {
        val me = currentMemberFromDb()
        val myApiKey = me.apiKey

        val latestSummaries = chatPersistencePort.findLatestChatSummariesByMember(myApiKey, me.id)
        if (latestSummaries.isEmpty()) {
            return RsData("200-1", "채팅 목록 조회 성공", emptyList())
        }

        val latestChats = chatPersistencePort.findChatsWithRoomsByIds(latestSummaries.map { it.latestChatId })
        if (latestChats.isEmpty()) {
            return RsData("200-1", "채팅 목록 조회 성공", emptyList())
        }

        val unreadCountMap = latestSummaries.associate { it.roomId to it.unreadCount }
        val opponentApiKeys = latestChats.mapNotNull { chat ->
            val room = chat.chatRoom ?: return@mapNotNull null
            if (room.sellerApiKey == myApiKey) room.buyerApiKey else room.sellerApiKey
        }.toSet()

        val opponentMap = chatMemberPort.findMembersByApiKeys(opponentApiKeys)

        val responseList = latestChats.mapNotNull { chat ->
            val room = chat.chatRoom ?: return@mapNotNull null
            val opponentKey = if (room.sellerApiKey == myApiKey) room.buyerApiKey else room.sellerApiKey
            val opponent = opponentMap[opponentKey] ?: return@mapNotNull null

            ChatRoomListResponse(
                roomId = room.roomId,
                opponentId = opponent.id,
                opponentNickname = opponent.nickname,
                opponentProfileImageUrl = opponent.profileImageUrl,
                opponentReputation = opponent.reputationScore ?: 50.0,
                lastMessage = chat.message,
                lastMessageDate = chat.createDate,
                unreadCount = unreadCountMap[room.roomId] ?: 0,
                itemId = room.itemId,
                itemName = room.itemName,
                itemImageUrl = room.itemImageUrl,
                itemPrice = room.itemPrice,
                txType = room.txType,
            )
        }.sortedByDescending { it.lastMessageDate }

        log.debug("채팅 목록 조회 완료 - 사용자: {}, 조회된 방 개수: {}", me.nickname, responseList.size)
        return RsData("200-1", "채팅 목록 조회 성공", responseList)
    }

    @Transactional
    override fun exitChatRoom(roomId: String): RsData<Void?> {
        val room = findActiveRoom(roomId)
        val me = currentMemberFromDb()

        requireChatParticipant(room, me.apiKey)
        room.exit(me.apiKey)

        if (room.isBothExited) {
            room.softDelete()
        }

        return RsData("200-1", "채팅방에서 퇴장하였습니다.", null)
    }

    private fun sendUserNotification(
        type: String,
        room: ChatRoom,
        recipient: ChatMemberInfo,
        opponent: ChatMemberInfo,
        message: String?,
        messageDate: LocalDateTime?,
    ) {
        runMessagingSafely(
            onSuccess = { log.info("개인 알림 전송 - Type: {}, Recipient: {}, RoomID: {}", type, recipient.id, room.roomId) },
            onFailure = { e -> log.error("개인 알림 전송 실패 - Recipient: {}, Error: {}", recipient.id, e.message) },
        ) {
            val unreadCount = when (type) {
                "NEW_ROOM" -> 0
                else -> chatPersistencePort.countUnreadMessagesByRoomId(room.roomId, recipient.id)
            }

            val notification = ChatNotification(
                type = type,
                roomId = room.roomId,
                opponentId = opponent.id,
                opponentNickname = opponent.nickname,
                opponentProfileImageUrl = opponent.profileImageUrl,
                lastMessage = message,
                lastMessageDate = messageDate,
                unreadCount = unreadCount,
                itemId = room.itemId,
                itemName = room.itemName,
                itemImageUrl = room.itemImageUrl,
                itemPrice = room.itemPrice,
                txType = room.txType,
            )

            chatPublishPort.publishUserNotification(recipient.id, notification)
        }
    }

    private inline fun runMessagingSafely(
        crossinline onSuccess: () -> Unit,
        crossinline onFailure: (Throwable) -> Unit,
        block: () -> Unit,
    ) {
        runCatching(block)
            .onSuccess { onSuccess() }
            .onFailure(onFailure)
    }

    private fun currentMemberFromDb(): ChatMemberInfo {
        val actor = rq.actor ?: throw ServiceException("401-1", "로그인이 필요합니다.")
        val now = System.currentTimeMillis()

        memberCache[actor.id]
            ?.takeIf { it.expiresAtMillis > now }
            ?.let { return it.member }

        return chatMemberPort.getMemberOrThrow(actor.id).also { fetched ->
            memberCache[actor.id] = CachedCurrentMember(
                member = fetched,
                expiresAtMillis = now + CURRENT_MEMBER_CACHE_TTL_MILLIS,
            )
        }
    }

    private fun findActiveRoom(roomId: String): ChatRoom =
        chatPersistencePort.findActiveRoom(roomId)
            ?: throw ServiceException("404-1", "존재하지 않는 채팅방입니다.")

    private fun requireChatParticipant(room: ChatRoom, apiKey: String) {
        if (room.sellerApiKey != apiKey && room.buyerApiKey != apiKey) {
            throw ServiceException("403-1", "해당 채팅방에 접근 권한이 없습니다.")
        }
    }

    private fun ensureNotSelfTrade(sellerApiKey: String, buyerApiKey: String) {
        if (sellerApiKey == buyerApiKey) {
            throw ServiceException("400-3", "본인의 상품에는 채팅을 개설할 수 없습니다.")
        }
    }

    private fun parseTxType(txType: String): ChatRoomType =
        try {
            ChatRoomType.valueOf(txType.uppercase(Locale.ROOT))
        } catch (_: IllegalArgumentException) {
            throw ServiceException("400-2", "잘못된 거래 유형입니다.")
        }

    private data class CreatedRoom(
        val room: ChatRoom,
        val sellerApiKey: String,
        val isNew: Boolean,
    )

    companion object {
        private val log = LoggerFactory.getLogger(ChatService::class.java)
        private const val CURRENT_MEMBER_CACHE_TTL_MILLIS = 60_000L
    }

    private data class CachedCurrentMember(
        val member: ChatMemberInfo,
        val expiresAtMillis: Long,
    )
}
